### PR TITLE
Implement exprs

### PR DESCRIPTION
**Pull request by Mostafa-Atallah2020**
I have modified __dev__ and __pow__ methods and class description of HOBOExpr class. Also, I have implemented the following classes HOBOVar, IntVar, SpinVar, and BinVar .

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -1,36 +0,0 @@
-default_language_version:
```
Comment (jplewa) [inline .pre-commit-config.yaml]: Please don't remove this file. To disable pre-commit locally, you can just run something like `pre-commit uninstall`

**Review by jplewa — CHANGES_REQUESTED** [review]
Please update your commit messages to follow the pattern we agreed on. Also, please merge the changes from develop into this branch (or rebase your branch onto develop - whichever approach works). If you don't, GitHub Actions won't get triggered for this PR.

**Comment by jplewa** [comment]
@Mostafa-Atallah2020 I see that you didn't update your old commit messages to the format we agreed on and that the new messages do not follow this format either (should be "Fix pylint issues" instead of "fix pylint issues"). I'd prefer to have the commits cleaned up before merging, but otherwise this PR will definitely need to have all the commits squashed into one when you merge it.

**Comment by jplewa** [comment]
Please add some dummy test to the project, because CI/CD is failing due to `CoverageWarning: No data was collected`. This can't be merged in its current state.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -33,4 +33,4 @@ repos:
       - id: pylint
         args: [ "./**/*.py" ]
         additional_dependencies:
-          - pyenchant
```
Comment (jplewa) [inline .pre-commit-config.yaml:36]: Please revert this change

File: hokrodelopt/expr.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
+import sympy as sm
+
+
+class HOBOExpr:
+    """
+    It construct expressions using sympy.
```
Comment (jplewa) [inline hokrodelopt/expr.py]: construct -> constructs

File: hokrodelopt/expr.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+            if char.isspace():
+                raise ValueError("a variable name shouldn't contain any spaces in it.")
+
+    @staticmethod
+    def __is_valid_var_lim(val):
+        if not (val is None or isinstance(val, (int, float))):
+            raise ValueError("variable limits must be numerical or None.")
+
+    def __init__(self, name, lb, ub):
+        HOBOExpr.__init__(self)
```
Comment (jplewa) [inline hokrodelopt/expr.py]: `HOBOExpr.__init__` has an `expr` parameter?

File: hokrodelopt/expr.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+    @ub.setter
+    def ub(self, val):
+        self.__is_valid_var_lim(val)
+        self.ub = val
+
+    def set_lims(self, lb, ub):
+        if lb is not None and ub is not None and lb > ub:
+            raise ValueError(
+                "The lower bound {} can't be greater than the ubber bound {}.".format(lb, ub)
```
Comment (jplewa) [inline hokrodelopt/expr.py]: ubber -> upper

File: README.md [hunk 6]
```diff
@@ -61,7 +61,7 @@ find and fix the errors.
 |-------------|------------------|--------------------|----------------|
 | `isort`     | `pyproject.toml` | `isort . --check`  | `isort .`      |
 | `black`     | `pyproject.toml` | `black . --check`  | `black .`      |
-| `flake8`    | `.flake8`        | `flake8`           | none           |
+| `flake8`    | `.flake8`        | `flake8 ./**/*.py` | none           |
```
Comment (jplewa) [inline README.md:64]: obraz

Comment (Mostafa-Atallah2020) [reply README.md:64]: using `flake8` only checks every single file in the project which is not required and takes a lot of time. @jplewa

File: .pylintrc [hunk 7]
```diff
@@ -7,4 +7,4 @@ enable=spelling,
 
 [SPELLING]
 spelling-dict=en_US
-spelling-ignore-words=hokrodelopt
+spelling-ignore-words=hokrodelopt, sympy
```
Comment (adamglos92) [inline .pylintrc]: There should be a way to create a text file in which the ignored words are listed one per line. Probably it is `wordlist.txt`

File: README.md (hunk 6 above)
Comment (adamglos92) [reply README.md:64]: But flake8 is important. it detects quite important mistakes. I would even say it may be the most important parser of all we have

File: hokrodelopt/expr.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+        """for expressions of the form (var1 * var2) from right"""
+        return self.__mul__(other)
+
+    def __sub__(self, other):
+        """for expressions of the form (var1 - var2)"""
+        return self.__add__(-other)
+
+    def __pow__(self, order):
+        """for expressions of the form (base ** exponent)"""
+        if isinstance(self, int) and isinstance(order, int) and order >= 0:
```
Comment (adamglos92) [inline hokrodelopt/expr.py]: here order can be arbitrary, right?

File: hokrodelopt/expr.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+            return sm.Integer(self) ** sm.Integer(order)
+        elif isinstance(self, int) and isinstance(order, int) and order < 0:
```
Comment (adamglos92) [inline hokrodelopt/expr.py]: This can be hidden within the expression from 41 if we remove the `order >= 0` condition

File: hokrodelopt/expr.py [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+    def __rmul__(self, other):
+        """for expressions of the form (var1 * var2) from right"""
+        return self.__mul__(other)
+
+    def __sub__(self, other):
+        """for expressions of the form (var1 - var2)"""
+        return self.__add__(-other)
+
+    def __pow__(self, order):
```
Comment (adamglos92) [inline hokrodelopt/expr.py]: we can just do assertion if order is an `int`

File: hokrodelopt/expr.py [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+        elif isinstance(self, int) and isinstance(order, int) and order < 0:
+            return sm.Integer(1) / (sm.Integer(self) ** sm.Integer(abs(order)))
+        elif isinstance(order, int) and order >= 1:
+            return HOBOExpr(self.expr ** sm.Integer(order))
+        elif isinstance(order, int) and order == 0:
+            return sm.Integer(1)
+        else:
+            raise TypeError("order cannot be negative.")
+
+    def __div__(self, other):
```
Comment (adamglos92) [inline hokrodelopt/expr.py:50]: I think it would be better if `other` would be asserted to be a number, and then just divided. We may have precision issues otherwise

File: hokrodelopt/expr.py [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+        """for expressions of the form (var1 / var2)"""
+        return HOBOExpr(self.expr / sm.Float(other))
+
+    def __neg__(self):
```
Comment (adamglos92) [inline hokrodelopt/expr.py:57]: This is corrected but seems to be overcomplicated. Why not `-self.expr` ? In fact similar comment applies to many places

File: hokrodelopt/expr.py [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+    Examples
+    --------
+    >>> HOBOVar('x', lb=-10, ub=10)
+    '-10 <= x <= 10'
+    """
+
+    @staticmethod
+    def __is_valid_var_name(name):
+        if len(name) < 1:
+            raise ValueError("a varible name shouldn't be empty.")
```
Comment (adamglos92) [inline hokrodelopt/expr.py:77]: typo (variable). Also `len(name) == 0` would be better, or even `assert len(name) > 0, "[error message]"`

File: hokrodelopt/expr.py [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+
+        for char in name:
+            if char.isspace():
```
Comment (adamglos92) [inline hokrodelopt/expr.py:80]: In general it is better to assert restricted than forbid. So we could simply say that all should be letters or numbers (is `.` or `-` or `;` or `漢`? We should also take into account sympy limitations, perhaps this will be sufficient?

File: hokrodelopt/expr.py [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+                raise ValueError("a variable name shouldn't contain any spaces in it.")
+
+    @staticmethod
+    def __is_valid_var_lim(val):
```
Comment (adamglos92) [inline hokrodelopt/expr.py:84]: I'm not sure if this function is good, this will need to be specified for each subclass anyway (Int class should not accept `float`)

File: hokrodelopt/expr.py [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+        if not (val is None or isinstance(val, (int, float))):
+            raise ValueError("variable limits must be numerical or None.")
+
+    def __init__(self, name, expr, lb, ub):
```
Comment (adamglos92) [inline hokrodelopt/expr.py:95]: I think `__init__` should be the first method listed (but I'm flexible here). But for sure `self.lb < self.ub` should be asserted, and `name` tested

File: hokrodelopt/expr.py [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+            lb_str = ""
+        if self.ub is not None:
+            ub_str = " <= " + str(self.ub)
+        else:
+            ub_str = ""
+        return "".join((lb_str, super(HOBOVar, self).__str__(), ub_str))
+
+
+class BinVar(HOBOVar):
+    def __init__(self, name, lb=None, ub=None):
```
Comment (adamglos92) [inline hokrodelopt/expr.py]: There should be no `lb` or `ub` in here at all

File: hokrodelopt/expr.py [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+        HOBOVar.__init__(self, name=name, lb=lb, ub=ub)
+        HOBOVar.set_lims(self, 0, 1)
+
+
+class SpinVar(HOBOVar):
+    def __init__(self, name, lb=None, ub=None):
```
Comment (adamglos92) [inline hokrodelopt/expr.py]: There should be no `lb` or `ub` at all

File: hokrodelopt/expr.py [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+        HOBOVar.__init__(self, name=name, lb=lb, ub=ub)
+        HOBOVar.set_lims(self, -1, 1)
+
+
+class IntVar(HOBOVar):
+    def __init__(self, name, ub, lb=None):
+        HOBOVar.__init__(self, name=name, lb=lb, ub=ub)
```
Comment (adamglos92) [inline hokrodelopt/expr.py]: This is overall strange. First, we create `IntVar` with bounds, then we set its limits. Why not do this in the first time?

File: README.md (hunk 6 above)
Comment (jplewa) [reply README.md:64]: Okay, makes sense

File: .pre-commit-config.yaml [hunk 20]
```diff
@@ -32,5 +32,4 @@ repos:
     hooks:
       - id: pylint
         args: [ "./**/*.py" ]
-        additional_dependencies:
-          - pyenchant
```
Comment (jplewa) [inline .pre-commit-config.yaml:36]: Could you explain this change to me?

Comment (Mostafa-Atallah2020) [reply .pre-commit-config.yaml:36]: oh sorry! my bad, I may have changed it unintentionally. everything should be back to normal now.

**Review by adamglos92 — CHANGES_REQUESTED** [review]

File: hokrodelopt/expr.py (hunk 8 above)
Comment (Mostafa-Atallah2020) [reply hokrodelopt/expr.py]: yes!

File: hokrodelopt/expr.py (hunk 9 above)
Comment (Mostafa-Atallah2020) [reply hokrodelopt/expr.py]: I don't understand could you make it clearer?

File: hokrodelopt/expr.py (hunk 10 above)
Comment (Mostafa-Atallah2020) [reply hokrodelopt/expr.py]: I think I already do that @adamglos92

File: hokrodelopt/expr.py (hunk 15 above)
Comment (Mostafa-Atallah2020) [reply hokrodelopt/expr.py:84]: i think spin and binary variables all have `int`  limits so no need for `float` limits here.

File: hokrodelopt/expr.py (hunk 8 above)
Comment (adamglos92) [reply hokrodelopt/expr.py]: I think we can safely assume for the moment order is int always. We don't really need to assert this in fact, instead we could replace definition as `__pow__(self, order:int)`

File: hokrodelopt/expr.py (hunk 10 above)
Comment (adamglos92) [reply hokrodelopt/expr.py]: Maybe we could just make `__pow__(self, order:int)` and skip the type assertion in if condition

File: hokrodelopt/expr.py (hunk 9 above)
Comment (adamglos92) [reply hokrodelopt/expr.py]: Actually I think we should forbid negative order. This hardly makes sense in our case

File: hokrodelopt/expr.py (hunk 15 above)
Comment (adamglos92) [reply hokrodelopt/expr.py:84]: But later we may have real bounded variables

File: .pre-commit-config.yaml [hunk 21]
```diff
@@ -33,4 +34,4 @@ repos:
       - id: pylint
         args: [ "./**/*.py" ]
         additional_dependencies:
-          - pyenchant
+          - pyenchant
```
Comment (jplewa) [inline .pre-commit-config.yaml]: Please add the newline you removed from the end of this file

File: hokrodelopt/expr.py (hunk 14 above)
Comment (Mostafa-Atallah2020) [reply hokrodelopt/expr.py:80]: space
chinese
dot

Comment (Mostafa-Atallah2020) [reply hokrodelopt/expr.py:80]: surprisingly, sympy can accept all of these special characters in its names even the spaces.

File: hokrodelopt/expr.py (hunk 15 above)
Comment (Mostafa-Atallah2020) [reply hokrodelopt/expr.py:84]: So can we make `int` for now or do I need to respecify it for each subclass separately?
